### ISSUE TITLE
Deprecate overlapping outputs

### ISSUE THREAD

**Issue by lptr** [gradle/gradle#28394]
Historically, Gradle allowed multiple tasks to write to the same directory. However, when tasks do this, it plays badly with many of our optimizations.

- If both tasks create file X, but with different content, the result of the build becomes dependent on which order we execute the tasks.
- If we run them in parallel, even more trouble can arise.
- When Gradle runs a task and finds some unknown output is already there, it's hard to provide meaningful error messages if overlapping outputs are allowed.
- Even if they write different kinds of files to the same directory, it makes figuring out what to store in the build cache (and how to load them) harder than necessary.
- And let's not even start talking about trying to auto-assign output locations to a pair of tasks like this. Let alone run them in a sandbox.

Therefore we want to move Gradle towards a future where overlapping outputs are a thing of the past. If a build is using them today, the user has a number of options:

- Go from a push-style/imperative build, where tasks are executed to put its output in the right place (think of a WAR file structure), turn the build into pull-style/declarative, where each task does its own thing without regards to how it fits into some final layout. In many cases there is no need to arrange the files on disk "just right" anyway, as e.g. a `War` task can take its inputs from several different sources.
- If a specific layout must be achieved on disk, use a `Copy` task to pull things together.
- If for some reason this is not possible, keep the tasks as they are, but mark them as `@Untracked`. This way they are allowed to do whatever, but will lose all optimizations, including parallel execution and caching/up-to-date checks. Gradle will also not check for stale outputs for such tasks.

**Comment by lptr** [comment]
Forbidding overlapping outputs will help identify problems like https://github.com/gradle/gradle/issues/27586, where two tasks are accidentally configured to produce the exact same output.

**Comment by iseki0** [comment]
But how to solve the problem on #2137 @lptr. And why you create a new thread?

**Comment by lptr** [comment]
We should add first-class support for bytecode instrumentation. I opened a new issue because this is not about investigation anymore, but about going ahead with forbidding overlapping outputs. (I could have renamed the old issue, too, but I did not see it before I created this one; sorry.)
